sunxi-boot0.c

```diff
@@ -10,218 +11,640 @@
+	fprintf(stream, "\t.header_size = 0x%08X,\n", header->header_size);
+	fprintf(stream, "\t.header_version = \"%c%c%c%c\",\n",
+		header->header_version[0], header->header_version[1],
+		header->header_version[2], header->header_version[3]);
+	fprintf(stream, "\t.return_address = 0x%08X,\n",
+		header->return_address);
+	fprintf(stream, "\t.run_address = 0x%08X,\n", header->run_address);
+	fprintf(stream, "\t.eGON_version = \"%c%c%c%c\",\n",
+		header->eGON_version[0], header->eGON_version[1],
+		header->eGON_version[2], header->eGON_version[3]);
+	fprintf(stream, "\t.platform_info = \"%c%c%c%c%c%c%c%c\",\n",
+		header->platform_info[0], header->platform_info[1],
+		header->platform_info[2], header->platform_info[3],
+		header->platform_info[4], header->platform_info[5],
+		header->platform_info[6], header->platform_info[7]);
+
+	fprintf(stream, "};\n\n");
+}
+
+static int
+egon_checksum_verify(FILE *stream, struct egon_header *header, FILE *inf)
+{
+#define BUFFER_COUNT (BOOT0_FILESIZE_ALIGN / sizeof(uint32_t))
+	uint32_t buffer[BUFFER_COUNT];
+	uint32_t checksum = EGON_CHECKSUM_SEED;
+	off_t checksum_offset =
+		offsetof(struct egon_header, checksum) / sizeof(uint32_t);
+	off_t offset = 0;
+
+	while (1) {
+		int ret, i;
+
+		ret = fread(buffer, sizeof(uint32_t), BUFFER_COUNT, inf);
+		if (ret != BUFFER_COUNT) {
+			fprintf(stream, "Error: %s(): fread failed: %s (%d)\n",
+				__func__, strerror(errno), ret);
+			return ret;
+		}
+
+		for (i = 0; i < BUFFER_COUNT; i++) {
+			if (!offset && (i == checksum_offset))
+				continue;
+			checksum += buffer[i];
+		}
+
+		offset += ret * sizeof(uint32_t);
+		if (offset == header->filesize)
+			break;
+	}
+
+	if (checksum != header->checksum)
+		fprintf(stream, "eGON checksum mismatch: 0x%08X vs 0x%08X\n",
+			checksum, header->checksum);
+	else
+		fprintf(stream, "eGON checksum matches.\n");
+
+	return 0;
+}
+
+struct egon_header_secondary {
+	uint32_t header_size;
+	char header_version[4];
+#define BOOT0_DRAM_PARAM_COUNT 32
+	uint32_t dram_param[BOOT0_DRAM_PARAM_COUNT];
+	/* ignore the rest of this struct for now. */
 };
 
-enum soc_types {
-	SOC_A64 = 0,
-	SOC_H616,
-	SOC_H6,
-	NR_SOC_TYPES
+static void __maybe_unused
+egon_header_secondary_print(FILE *stream,
+			    struct egon_header_secondary *header)
+{
+	int i;
+
+	fprintf(stream, "struct egon_header header[1] = {\n");
+
+	fprintf(stream, "\t.header_size = 0x%08X,\n", header->header_size);
+	fprintf(stream, "\t.header_version = \"%c%c%c%c\",\n",
+	       header->header_version[0],
+	       header->header_version[1],
+	       header->header_version[2],
+	       header->header_version[3]);
+
+	for (i = 0; i < BOOT0_DRAM_PARAM_COUNT; i++)
+		fprintf(stream, "\t.dram_param[0x%02X] = 0x%08X,\n",
+		       i, header->dram_param[i]);
+
+	fprintf(stream, "\t/* ... */\n");
+
+	fprintf(stream, "};\n\n");
+}
+
+/*
+ * should work for A10, A10s, A13 and A20.
+ */
+#define DRAM_PARAM_A10_MATCHES "A10/A10s/A13/A20"
+
+struct dram_param_a10 {
```

Comment:
The sed/awk-crafted printfs allow for extra comments useful to the user. cfr the h616 printing.

This is also not per SoC, it is per group of SoCs for which this dram para struct was commonly seen. All in all, we have 4.

Since i am verifying some values to figure out which structure is likely relevant, just referencing struct members makes it easy and very readable.